### ISSUE TITLE
PCWSTR arguments/fields are generated as PWSTR in in Rust bindings

### ISSUE THREAD

**Issue by randomPoison** [microsoft/windows-rs#777]
I'm currently using the `windows` crate to generate bindings the the Projected Filesystem API, and I noticed that parts of the API that use `PCWSTR` (both function arguments and struct fields) are being treated as `PWSTR` (i.e. they mutable wide strings instead of const wide strings).

A simple example of this is the [`PRJ_GET_DIRECTORY_ENUMERATION_CB`](https://docs.microsoft.com/en-us/windows/win32/api/projectedfslib/nc-projectedfslib-prj_get_directory_enumeration_cb) type, which is generated as follows with `windows` 0.9:

```rust
pub type PRJ_GET_DIRECTORY_ENUMERATION_CB =
    unsafe extern "system" fn(
        callbackdata: *const PRJ_CALLBACK_DATA,
        enumerationid: *const ::windows::Guid,
        searchexpression: super::SystemServices::PWSTR,
        direntrybufferhandle: PRJ_DIR_ENTRY_BUFFER_HANDLE,
    ) -> ::windows::HRESULT;
```

Note that the `searchExpression` argument is declared as a `PWSTR` in the generated Rust bindings, but the original definition declares it as a `PCWSTR`.

To reproduce this, you can create a build script that generates the ProjFS bindings:

```rust
windows::build!(
    Windows::Win32::ProjectedFileSystem::*,
);
```

**Comment by ennis** [comment]
Bindings to DirectWrite have the same issue: for example, [IDWriteFactory::CreateTextLayout](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-idwritefactory-createtextlayout), which takes a `const WCHAR*`:
```
HRESULT CreateTextLayout(
  WCHAR const       *string,
  UINT32            stringLength,
  IDWriteTextFormat *textFormat,
  FLOAT             maxWidth,
  FLOAT             maxHeight,
  IDWriteTextLayout **textLayout
);
```

gets translated into a function that takes a PWSTR (`*mut u16`):
```
pub unsafe fn CreateTextLayout<'a>(
                    &self,
                    string: impl ::windows::IntoParam<'a, super::SystemServices::PWSTR>,
                    stringlength: u32,
                    textformat: impl ::windows::IntoParam<'a, IDWriteTextFormat>,
                    maxwidth: f32,
                    maxheight: f32,
                    textlayout: *mut ::std::option::Option<IDWriteTextLayout>,
                ) -> ::windows::HRESULT { ... }
```

**Comment by kennykerr** [comment]
Sorry for the delay in responding. The metadata that drives the `Windows` crate only defined `PSTR` and `PWSTR` - there's no `const` version. The `const` is translated into an attribute, either on the parameter or the field, that languages can then interpret as needed. The `Windows` crate already takes into account the `const` attribute on function parameters, but not structs or callbacks. That still remains to be mapped in some sensible way. I have some work coming that will deal with this more naturally, both `const` and `optional` (#816) attributes.

**Comment by lzybkr** [comment]
@kennykerr - I found at least 1 parameter that should be `const` but is not: `lpmultibytestr` on `MultiByteToWideChar`:

```rust
    pub fn MultiByteToWideChar(codepage: u32, dwflags: MULTI_BYTE_TO_WIDE_CHAR_FLAGS, lpmultibytestr: super::Foundation::PSTR, cbmultibyte: i32, lpwidecharstr: super::Foundation::PWSTR, cchwidechar: i32) -> i32;
```

Version:

```toml
[dependencies.windows-sys]
version = "0.27.0"
default-features = false
features = ["Win32_Globalization", "Win32_Foundation"]
```

**Comment by kennykerr** [comment]
@lzybkr thanks for the reminder - I should be able to generate `const` in such cases.

**Comment by kennykerr** [comment]
This was recently fixed with #1468 and #1470
